---
title: "Meningitis_practice"
author: "Geraint Sunderland"
date: "`r format(Sys.time(), '%d/%m/%Y')`"
output: pdf_document
keep_md: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(root.dir = '/Users/gjs12782/iCloudDrive/PhD/Oxford proteomics/Data_Analysis/Analysis_tidy/Meningitis_practice/')
knitr::opts_chunk$set(dev = "jpeg",
                      dpi = 300,
                      echo = FALSE,
                      eval = TRUE,
                      message = FALSE,
                      warning = FALSE,
                      cache = TRUE)
knitr::write_bib(c(.packages(), "bookdown"), "packages.bib")
```

```{r}

library(EnsDb.Hsapiens.v79)
library(tidyverse)
library(org.Hs.eg.db)
library(preprocessCore)
library(cluster)
library(factoextra)
library(gplots)
library(ggplot2)
library(RColorBrewer)
library(ggrepel)
library(edgeR)
library(limma)
library(qvalue)
library(effsize)
library(tmaptools)
library(pheatmap)
library(reshape2)
library(biomaRt)
library(clusterProfiler)
library(vsn)
library(e1071)


```


```{r}
meningitis_data <- read.csv('meningitis_data_final.csv')
head(meningitis_data[1:5, 1:5])
```




Create phenotype data frame for later

```{r}
Group <- as.data.frame(cbind(c(rep(1, 16), rep(2, 16)), c(rep('BM', 16), rep('NAD', 16))))
colnames(Group)<-c('Group', 'BM_NAD')
# head(Group)
```


```{r}

meningitis_data1 <- meningitis_data[,3:ncol(meningitis_data)]
rownames(meningitis_data1)<- paste(meningitis_data[,2])
#Transpose data                   
tmeningitis_data1 <- t(meningitis_data1)


tmeningitis_master_data <- as.data.frame(t(meningitis_data[,3:ncol(meningitis_data)]))
colnames(tmeningitis_master_data)<- t(meningitis_data[,2])


tmeningitis_master_data1<-lapply(tmeningitis_master_data,as.numeric)
tmeningitis_master_data1 <- (cbind(Group, tmeningitis_master_data))
```

##Data Visualisation
First plot meanSD and barplot to look at data distribution/normality

```{r unnormalised_meanSDplot}

#Check the distribution of abundance values for each gel and any dependence of the
#standard deviation with the mean:

meningitis_data_mat <- as.matrix(meningitis_data[,4:ncol(meningitis_data)])

meanSdPlot(meningitis_data_mat)


```

```{r unnormalised_boxplot}
boxplot_cols <- c(rep("#F46D43",16), rep("#74ADD1", 16))

boxplot(t(tmeningitis_data1), main = "Unnormalised", outline = FALSE, col = boxplot_cols)
legend("topleft", legend=c("BM", "NAD"),fill=c("#F46D43", "#74ADD1"))

```

Normalise with vsn (Variance Stabilising Normalisation)

```{r vsn_normalised_meanSDplot}
meningitis_norm.data1 <- t(justvsn(as.matrix(meningitis_data1)))
# library(proDA)
# meningitis_norm.data1 <- t(median_normalization(as.matrix(meningitis_data1)))

#View distribution of mean vs sd
meanSdPlot(t(meningitis_norm.data1))
```
      
```{r vsn_normalised_boxplot}
boxplot(t(meningitis_norm.data1), main = "Normalised", outline=FALSE, col = boxplot_cols)
legend("topleft", legend=c("BM", "NAD"),fill=c("#F46D43", "#74ADD1"))

```

Exploratory analysis
1. PCA

```{r }
meningitis_data1_PCA <- prcomp(as.matrix(tmeningitis_master_data1[,3:ncol(tmeningitis_master_data1)]), center = T, scale. = F)
# summary(meningitis_data1_PCA)
```

Create object to demonstrate relative loadings of PCs

```{r }
meningitis_data1_varExplained <- summary(meningitis_data1_PCA)$importance[2,]*100

meningitis_data1_varExplained
```

Make data frame containing PC1 and PC2 scores (PC1 and PC2 explain the vast majority of variance)

```{r }
meningitis_data1PCA_score1_2 <- as.data.frame(meningitis_data1_PCA$x[,1:2])
```

Plot PCA:

```{r meningitis_PCA}
library(ggplot2)
ggplot(meningitis_data1PCA_score1_2,aes(x=PC1,y=PC2))+
  geom_point(size=3,aes(col=tmeningitis_master_data1$Group))+#attaches the case type Infected/Clean
  scale_color_manual(values = c("1" = "#F46D43", "2" = "#74ADD1"), labels=c("BM", "NAD"), name = "Group")+ 
  xlab(paste("PC1 ","(",meningitis_data1_varExplained[1],"%)",sep=""))+
  ylab(paste("PC2 ","(",meningitis_data1_varExplained[2],"%)",sep=""))+
  theme(legend.position = "right")
```

Can look to see what genes/peptides are contributing to the loadings of the PCs using a loadings plot

```{r meningitis_PCA_loadings}
meningitis_data1_loadings1_2 <- as.data.frame(meningitis_data1_PCA$rotation[, 1:2])

ggplot(meningitis_data1_loadings1_2,aes(x=PC1,y=PC2))+
  geom_text(aes(x=PC1,y=PC2),label=rownames(meningitis_data1_loadings1_2), size=3)+
  xlab("Loadings of PC1")+
  ylab("Loadings of PC2")+
  theme_bw(base_size = 12)
```

# Heatmap/cluster analysis

Use elbow plot analysis to indicate how many clusters to explore

```{r meningitis_elbow}
wss <- (nrow(meningitis_norm.data1) - 1) * sum(apply(meningitis_norm.data1, 2, var))
for (i in 2:20) wss[i] <- sum(kmeans(meningitis_norm.data1, centers = i)$withinss)

plot(1:20, wss, type = "b", xlab = "Number of Clusters", ylab = "Within groups sum of squares")
```

Alternate method:

```{r meningitis_elbow2}
KmeansResult <- kmeans(x = meningitis_norm.data1, centers = 2)
KmeansResult$cluster



#elbow plot to evaluate optimal number of clusters
fviz_nbclust(meningitis_norm.data1, kmeans, method = "wss")
```

```{r menigitis_silhouette}
fviz_nbclust(meningitis_norm.data1, kmeans, method = "silhouette")
```

#kmeans cluster

```{r meningitis_kmeans_cluster}
final <- kmeans(meningitis_norm.data1, 2, nstart = 25)
fviz_cluster(final, data = meningitis_norm.data1, labelsize = 1)+
  scale_fill_manual(name = "",
                    labels = c("Cluster 1", "Cluster 2"),
                     values = c("#F46D43", "#74ADD1"))+
  scale_shape_manual(name = "",
                     labels = c("Cluster 1", "Cluster 2"),
                     values = c(16, 16))+
  scale_color_manual(name = "",
                     labels = c("Cluster 1", "Cluster 2"),
                     values = c("#F46D43", "#74ADD1"))+
  ggtitle("")
```

Clustering

2 way clustering:

```{r meningitis_2way_cluster}

hr <- hclust(as.dist(1-cor(meningitis_norm.data1, method = "pearson")), method = "average")
hc <- hclust(as.dist(1-cor(t(meningitis_norm.data1), method = "pearson")), method = "average")
mycl <- cutree(hr, h = max(hr$height)/1.008) 

mycolhc <- rainbow(length(unique(mycl)), start = 0.1, end = 0.9); mycolhc  <- mycolhc[as.vector(mycl)]

mycol <- colorpanel(30, "#4575B4", "grey95", "#D73027")

heatmap.2(t(meningitis_norm.data1), 
          RowSideColors = mycolhc, 
          Rowv = as.dendrogram(hr), 
          Colv = as.dendrogram(hc), 
          col = mycol,
          density.info = "none", 
          trace = "none", 
          dendrogram = "both", 
          scale = "row",
          hclustfun=function(d) hclust(d,method="average"),
          labRow = TRUE)
```


#One way clustering

```{r meningitis_1way_cluster}

heatmap.2(t(meningitis_norm.data1), 
          RowSideColors = mycolhc, 
          Rowv = as.dendrogram(hr), 
          Colv = FALSE, 
          col = mycol,
          density.info = "none", 
          trace = "none", 
          dendrogram = "both", 
          scale = "row",
          hclustfun=function(d) hclust(d,method="average"),
          labRow = TRUE)

# pheatmap(t(meningitis_norm.data1),color = colorRampPalette(c("#4575B4", "grey97", "#D73027"))(30),cluster_rows = TRUE, cluster_cols = FALSE, clustering_distance_rows = "correlation", clustering_distance_columns = "correlation", scale="row",show_rownames=FALSE, clustering_method = "average")
```

#Data analysis
First step should be univariate hypothesis testing with multiple comparison correction - this example using Smyth's moderated t-test with Storey's q-value method to control the FDR. Size of effect reported typically as fold change or in this case cohen's D

```{r }

design.mtrx <- cbind(rep(1,length(Group)), (as.matrix(as.numeric(Group$Group))))
data.transposed <- as.matrix(t(meningitis_norm.data1))
fit <- lmFit(data.transposed,design=design.mtrx,method="robust",maxit=1024)
fit <- eBayes(fit,robust=TRUE)
qval <- (qvalue(fit$p.value[,2],pi0.method="bootstrap"))$qvalues
fx.size <- apply(meningitis_norm.data1,2,
                 function(d,f) cohen.d(d,f)$estimate,f=factor(Group$Group))
sig.peptides <- names(qval[qval < 0.1]) # FDR < 10%
n.sig.peptides <- length(sig.peptides) # number of significant peptides
n.sig.peptides
peptide.class <- as.numeric(colnames(meningitis_norm.data1) %in% sig.peptides)
```

t-test assumes that errors are normally distributed. Check if this assumption is correct by looking at distribution of the residuals. A few options are to either studentize the per-peptide residuals, pool them all and compare the resulting distribution with a normal distribution via a Q-Q plot or to look at the skewness and kurtosis distributions of the per-peptide residuals:

```{r meningitis_QQplot}
# pooled studentized residuals approach
fit.residuals <- residuals(fit,data.transposed)
fit.residuals.student <- as.vector(scale(fit.residuals))
qqnorm(fit.residuals.student)
abline(0,1)
```

# skewness/kurtosis approach

```{r meningitis_skew_kurtosis}
set.seed(1) # make it reproducible
skew2 <- function(x) skewness(x,type=2)
kurt2 <- function(x) kurtosis(x,type=2)
normal.variates <-
  matrix(0,ncol=ncol(fit.residuals),nrow=nrow(fit.residuals))
for (i in 1:nrow(fit.residuals)) normal.variates[i,] <-
  rnorm(ncol(fit.residuals),0,1)
normal.skew <- apply(normal.variates,1,skew2)
normal.kurt <- apply(normal.variates,1,kurt2)
residual.skew <- apply(fit.residuals,1,skew2)
residual.kurt <- apply(fit.residuals,1,kurt2)
plot(NA,xlab="Skewness",ylab="Excess kurtosis",
     xlim=c(min(c(normal.skew,residual.skew)),max(c(normal.skew,residual.skew))),
     ylim=c(min(c(normal.kurt,residual.kurt)),max(c(normal.kurt,residual.kurt))))
points(normal.skew,normal.kurt,pch=".",cex=5,col="grey")
abline(v=mean(normal.skew),col="grey")
abline(h=mean(normal.kurt),col="grey")
points(residual.skew,residual.kurt,pch=".",cex=5)
abline(v=mean(residual.skew))
abline(h=mean(residual.kurt))
```

Both plots suggest that the distribution of the residuals is approximately normal, which supports the application of statistical tests that assume errors to be normally distributed.

The results of univariate hypothesis testing can be visualized using a volcano plot:

```{r meningitis_basic_volcano}
peptide.colours <-
  as.numeric(colnames(meningitis_norm.data1) %in% names(qval[qval < 0.05])) +
  1 + peptide.class
plot(fx.size,-log(qval)/log(10),col=peptide.colours,
     xlab="Effect size (Cohen's d)",ylab="Significance (-log10(q-value))",
     pch=0.1,cex=0.3)
legend("top",c("FDR > 5%","FDR < 5%","FDR < 1%"),col=1:3,pch=15)
```

Compile results into dataframe and save as .csv

```{r}
univariate.results <-
  data.frame(peptide.name=colnames(meningitis_norm.data1),p.value=fit$p.value[,2],
             q.value=qval,effect.size=fx.size,significant=peptide.class)
# add a column of NAs
univariate.results$diffexpressed <- "NO"
# if CohenD > 0.6 and pvalue < 0.05, set as "UP" 
univariate.results$diffexpressed[univariate.results$effect.size > 0.5 & univariate.results$q.value < 0.05] <- "UP"
# if CohenD < -0.6 and pvalue < 0.05, set as "DOWN"
univariate.results$diffexpressed[univariate.results$effect.size < -0.5 & univariate.results$q.value < 0.05] <- "DOWN"
univariate.results$label <- NA
univariate.results$label[univariate.results$diffexpressed != "NO"] <- univariate.results$peptide.name[univariate.results$diffexpressed != "NO"]

write.csv(univariate.results,"univariate_results_filtered.csv")
```

Nicer volcano plot in ggplot

```{r meningitis_volcano}

vol_col <- c("#4575B4", "grey60", "#D73027")

ggplot(data=univariate.results, aes(x=effect.size, y=-log10(q.value), col=diffexpressed, label=label))+ 
  geom_point() + 
  scale_color_manual(values = c("#4575B4", "grey60", "#D73027"), name = "Expression Level",
                    labels = c("Down-regulated\n(q\u22640.05)\n", "Unchanged\n(q>0.05)\n", "Up-regulated\n(q\u22640.05)"))+
  geom_vline(xintercept=c(-0.6, 0.6), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red")+
  xlab("Effect Size (Cohen's d")


```

```{r meningitis_volcano_annotated}

top <- 20
top_genes <- bind_rows(
  univariate.results %>% 
    filter(diffexpressed == 'UP') %>% 
    arrange(desc(effect.size), desc(abs(q.value))) %>% 
    head(top),
  univariate.results %>% 
    filter(diffexpressed == 'DOWN') %>% 
    arrange(effect.size, desc(abs(q.value))) %>% 
    head(top))

ggplot(data=univariate.results, aes(x=effect.size, y=-log10(q.value), col=diffexpressed, label=label))+ 
  geom_point() + 
  scale_color_manual(values = c("#4575B4", "grey60", "#D73027"), name = "Expression Level",
                    labels = c("Down-regulated\n(q\u22640.05)\n", "Unchanged\n(q>0.05)\n", "Up-regulated\n(q\u22640.05)"))+
  geom_label_repel(data = top_genes,
                   mapping = aes(x=effect.size, y= -log10(q.value), label = peptide.name),
                   size = 2, show.legend = "none")+
  geom_vline(xintercept=c(-0.6, 0.6), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red")+
  xlab("Effect Size (Cohen's d")

```


Compile results into dataframe and save as .csv

```{r }
meningitis_univariate.results <-
  data.frame(peptide.name=colnames(meningitis_norm.data1),p.value=fit$p.value[,2],
             q.value=qval,effect.size=fx.size,significant=peptide.class)
write.csv(meningitis_univariate.results,"meningitis_filtered_univariate_results1.csv")
```

Need to generate/convert geneIDs from peptide symbol. 
First produce character vector of peptide symbols

```{r}
peptide_symbol <- base::as.vector(meningitis_data [,1])
peptide_symbol[1:5]

```

```{r}

geneIDs <- ensembldb::select(EnsDb.Hsapiens.v79, keys= UniprotFilter(peptide_symbol),
                              keytype = "SYMBOL", columns = c("SYMBOL","GENEID"))
#Remove LNC additional info/duplicate rows - to match no of rows in dataset
Unique_geneIDs <- geneIDs %>% distinct(SYMBOL, .keep_all = TRUE)
Unique_geneIDs <- dplyr::rename(Unique_geneIDs, 'Gene_Name'='SYMBOL')

meningitis_Age_Data <- meningitis_data %>% dplyr::rename(UNIPROTID = Protein_ID)

meningitis_data1 <- inner_join(Unique_geneIDs, meningitis_Age_Data)
meningitis_data <- meningitis_data1
# head(data[1:5, 1:5])
#Compile as .csv
write_csv(meningitis_data, 'meningitis_GENID.csv')
```

Create phenotype metadata dataframe for later

```{r }
  meningitis_phenotype <- read.csv('meningitis_phenotype_final.csv', row.names = 1, 
                      header = T)
# head(meningitis_phenotype)
```



```{r }

cols <- colorRampPalette(c("#4575B4", "#D73027"))(32)

```

Convert counts to a differentially expressed genes list 
(DGEList - EdgeR function)

```{r }
data_dge <- DGEList(meningitis_data[ ,4:ncol(meningitis_data)])
class(data_dge)
#1st element of DGEList is the gene counts
head(data_dge[["counts"]])
#assign geneIDs from Ensemble_ID column to rownames of counts/DGEList
rownames(data_dge[["counts"]]) <-meningitis_data$GENEID
#check assigned correctly
head(rownames(data_dge[["counts"]]))
```

2nd element in DGEList is samples object - a dataframe containing data on 
samples

```{r }
head(data_dge[["samples"]])
#automatically calculated Library size (no of reads per sample), also 
#assigned groups but this needs changing as this will determine comparisons made
#assign groups from phenotype table - first create a vector withe that 
#information
samp_groups <- as.factor(meningitis_phenotype$Group)
samp_groups
#Use this to reassign gropus in DGEList
data_dge[["samples"]]$group <- samp_groups
#check to see correctly assigned
# data_dge[["samples"]]
```

Next need to normalise data to correct for potential errors- calculates library 
size from count but this could be due to greater library depth rather than 
no of reads.
EdgeR does this by converting data to counts per million (CPM)
Looking at variability in the library size

```{r }
# data_dge[["samples"]]$lib.size
max(data_dge[["samples"]]$lib.size)/min(data_dge[["samples"]]$lib.size)
#reveals that max library size is 9.2x that of min - confounder 
```

Can access normalised data from data set by calling the following

```{r eval=TRUE, echo= TRUE}
data_cpm <- cpm(data_dge)
#logCPM
log_cpm <- cpm(data_dge, log = TRUE)
```

To remove genes with low expression - reduces the no of tests needed to perform and increases power to find differentially expressed genes after correction for multiple testing
First plot Density plot to demonstrate gene count distribution - between the bars are genes w low expression - (-ve values present due to function of EdgeR cpm(log=TRUE) which adds a small psuedocount/offset to genes to prevent taking the log of 0.)

```{r meningitis_count_density}
log_cpm_melt <- melt(log_cpm)

head(log_cpm_melt)

ggplot(log_cpm_melt, aes(x = value, col = Var2)) +
  geom_density() +
  scale_colour_manual(values = cols) +
  geom_vline(xintercept = c(-5,-3)) +
  labs(col = "Samples") +
  guides(col = guide_legend(override.aes = list(size = 5, shape = 15)))+
  theme(legend.position="right")
```

Use filterByExpr function in EdgeR to remove genes with low expression - not
relevant in this case but included 'for example'.

```{r }
genes_with_suff_counts <- filterByExpr(data_dge, group = 
                                               data_dge[["samples"]]$group)
sum(genes_with_suff_counts)
```

230 genes retained of 243
If needed you can now subset DGEList with

```{r} 
data_dge <- data_dge[genes_with_suff_counts,, keep.lib.sizes = FALSE]
``` 

not relevant in this case but then produce a 'filtered' dataset as below:

```{r meningitis_filtered_counts_density}
filtered_cpm <- cpm(data_dge)
filtered_log_cpm <- cpm(data_dge, log = TRUE)
#Replot density plot
ggplot(melt(filtered_log_cpm), aes(x = value, col = Var2)) +
  geom_density() +
  scale_colour_manual(values = cols) +
  geom_vline(xintercept = c(-5,-3)) +
  labs(col = "Samples") +
  guides(col = guide_legend(override.aes = list(size = 5, shape = 15)))+
  theme(legend.position="right")
```

Now data is filtered can perform further normalisation step - in EdgeR this is with
trimmed mean of M values (TMM)

```{r menignitis_logNormalised_boxplot}
boxplot(cpm(data_dge[["counts"]], log = T), col = boxplot_cols, main = "Log normalised", outline = FALSE)
legend("topleft", legend=c("BM", "NAD"),fill=c("#F46D43", "#74ADD1"))
```

```{r meningitis_TMMnormalised_boxplot}
norm_data_dge <- calcNormFactors(data_dge, method = "TMM")
norm_cpm <- cpm(norm_data_dge, normalized.lib.sizes = TRUE)
norm_log_cpm <- cpm(norm_data_dge, log = TRUE, normalized.lib.sizes = TRUE)

boxplot(norm_log_cpm, col = boxplot_cols, main = "TMM Normalised", outline = FALSE)
legend("topleft", legend=c("BM", "NAD"),fill=c("#F46D43", "#74ADD1"))




```

Testing for differential expression - test for statistical differences between
control and case samples - in this setting NAD and BM cases Set up a design matrix. 
By default the first factor is used as an intercept - i.e subsequent factors will 
be compared with this - to avoid this need to pass the call ~0 + group.
To compute more complex interactions this is required. In a simple case such as 
this one comparing 2 groups only this does not matter

```{r meningitis_voom_mean_variance_trend}
design <- model.matrix(~0 + group, data = data_dge[["samples"]])
#remove 'group' from colnames
colnames(design) <- gsub("group", "", colnames(design))
#Call voom function - transforms counts to log2cpm
voom_data <-voom(norm_data_dge, design, plot = TRUE)
#lmfit with compute linear models
voom_fit <- lmFit(object = voom_data, design = design)
#Identify the contrasts wanted - i.e BM vs NAD 'BMvsNAD' - the product of AD 
#expression values minus control expression values (BM-NAD) The outcome is that 
#anything with +ve fold change has higher expression in BM and -ve fold change 
#lower expression in BM.
cont.matrix <- makeContrasts(BMvsNAD = BM - NAD, levels = design)
#pass voom_fit object with cont.matrix
voom_fit <- contrasts.fit(fit = voom_fit, contrasts = cont.matrix)
#Use empirical Bayes method to reduce variances - based on variances calculated 
#from the dataset
voom_fit <- eBayes(voom_fit)
```

plotSA (sigma Vs Alpha)This plot is used to check the mean-variance relationship of the expression data, after fitting a linear model. A scatterplot of residual-variances vs average log-expression is created. The plot is especially useful for examining the mean-variance trend estimated by eBayes or treat with trend=TRUE. It can be considered as a routine diagnostic plot in the limma-trend pipeline.
#diagnostic plot to show mean variances after fitting linear model

```{r meningitis_SAplot}
plotSA(voom_fit)
```

Use topTable to display differentially expressed genes (DEGs).

```{r }
voom_tt <- topTable(voom_fit, number = Inf)
#Convert IDs to genes - using biomaRt - connects to annotation databases - need 
#to set up mart object and which particular organiNAD eg. Human
#Display available marts
listMarts()
#Select ensembl to view gene annotations
ensembl=useMart("ENSEMBL_MART_ENSEMBL")
list_of_datasets <- listDatasets(ensembl)
#use hsapiens dataset
ensembl <- useDataset("hsapiens_gene_ensembl", mart = ensembl)
#Browse list of filters - have ensembl ids so narrow down to ensembl options 
#using 'grep'
list_of_filters <- listFilters(ensembl)
list_of_filters[grep("entr", list_of_filters$names),]
filter_type <- "ensembl_gene_id"
#finally specify attributes to retrieve - give mart a filter containing all 
#Ensembl IDs we want it to return matched to gene names and descriptions
listAttributes(ensembl)[grep("", listAttributes(ensembl)),]
attribute_names <- c('ensembl_gene_id', 'external_gene_name', 'description')
filter_values <- rownames(voom_tt)
annot <- getBM(attributes=attribute_names,
               filters = filter_type,
               values = filter_values,
               mart = ensembl)
#Join topTable and annotation object together - convert rowname to column and 
#give it the variable Ensemble_id
#Use inner_join function to combine by matching "Ensembl_id" from tt to 
#"ensembl_gene_id" from annot
voom_tt <- voom_tt %>% rownames_to_column(var = "Ensembl_id")
voom_tt <- voom_tt %>% inner_join(annot, by = c("Ensembl_id" = "ensembl_gene_id"))
#Write to .csv file
write.csv(voom_tt, 'meningitis_summary_cluster_analysis_filtered.csv')
```

Plot heatmap of significant genes  that demonstrates clear separation between 
groups - example shows top 20 so easy to read but can show as many as you like

```{r meningitis_DEG_heatmap}

heatmap_signif_cpm <- norm_cpm[rownames(norm_cpm) %in% voom_tt$Ensembl_id,]
heatmap_signif_cpm <- heatmap_signif_cpm %>% as.data.frame() %>%
  rownames_to_column(var = "Ensembl_id") %>%
  inner_join(voom_tt[,c(1,8:9)]) %>%
  column_to_rownames(var = "external_gene_name")
mycol <- colorpanel(75, "#4575B4", "grey97", "#D73027")
heatmap.2(as.matrix(heatmap_signif_cpm[,-c(1,34)]),
          Rowv = T, Colv = T,
          trace="none",
          dendrogram = "column",
          col=mycol,
          margins=c(7,10),
          scale = "row", cexRow = 0.5,
          hclustfun=function(d) hclust(d,method="average"))
```
Show upregulated genes only


```{r meningitis_upreg_DEG_heatmap}
meningitis_Upreg <- dplyr::filter(voom_tt, logFC >= 1)


heatmap_signif_cpm <- norm_cpm[rownames(norm_cpm) %in% meningitis_Upreg$Ensembl_id,]
heatmap_signif_cpm <- heatmap_signif_cpm %>% as.data.frame() %>%
  rownames_to_column(var = "Ensembl_id") %>%
  inner_join(meningitis_Upreg[,c(1,8:9)]) %>%
  column_to_rownames(var = "external_gene_name")
heatmap.2(as.matrix(heatmap_signif_cpm[,-c(1,34)]),
          Rowv = T, Colv = T,
          trace="none",
          dendrogram = "column",
          col=mycol,
          margins=c(7,10),
          scale = "row", cexRow = 0.5,
          hclustfun=function(d) hclust(d,method="average"))
```

Use cluster profiler to get demonstration of the biological functions represented 
in the differentially expressed genes

```{r meningitis_GOtermsBP_DEG_barplot}

Cluster_GO <- enrichGO(voom_tt$Ensembl_id,
                       OrgDb = org.Hs.eg.db,
                       ont = "BP",
                       pvalueCutoff = 0.05,
                       keyType = "ENSEMBL",
                       readable = TRUE)

Cluster_upreg_GO <- enrichGO(meningitis_Upreg$Ensembl_id,
                                           OrgDb = org.Hs.eg.db,
                                          ont = 'BP',
                                           pvalueCutoff = 0.05,
                                           keyType = "ENSEMBL",
                                           readable = TRUE)



barplot(Cluster_GO, showCategory = 30, font.size = 8)
```

Upreg only

```{r meningitis_GOtermsBP_Upreg_DEG_barplot}
barplot(Cluster_upreg_GO, showCategory = 30, font.size = 8)

```


```{r meningitis_GOtermsBP_DEG_dotplot}
dotplot(Cluster_GO, showCategory = 30, font.size = 8)

# "The top 20 ranked GO terms according to gene count. ‘Qvalue’ is the P-value adjusted using the Benjamini-Hochberg procedure. ‘Log (Qvalue)’ is the log 10 of the q-value. ‘Gene count’ is the number of genes enriched in a GO term. ‘Gene ratio’ is the percentage of total DEGs in the given GO term (only input genes with at least one GO term annotation were included in the calculation). GO, gene ontology; BP, biological process; MF, molecular function; CC, cellular component."
```

Upreg

```{r meningitis_GOtermsBP_Upreg_DEG_dotplot}
dotplot(Cluster_upreg_GO, showCategory = 30, font.size = 8)
```


Using KEGG database:

```{r meningitis_KEGG_barplot}

voom_tt <- topTable(voom_fit, number = Inf)
attribute_names2 <- c("ensembl_gene_id", "external_gene_name", "description", "entrezgene_id")
filter_type <- "ensembl_gene_id"
filter_values <- rownames(voom_tt)
annot2 <- getBM(attributes=attribute_names2, filters = filter_type, values = filter_values, mart = ensembl)
voom_tt2 <- voom_tt %>% rownames_to_column(var = "Ensembl_id")
voom_tt2 <- voom_tt2 %>% inner_join(annot2, by = c("Ensembl_id" = "ensembl_gene_id"))
Upreg2 = filter(voom_tt2, logFC >= 1)

EnrichKEGG = enrichKEGG(voom_tt2$entrezgene_id, organism = "hsa",keyType = "ncbi-geneid", qvalueCutoff = 0.05, use_internal_data = TRUE)

barplot(EnrichKEGG, showCategory = 30)
```

Upreg

```{r meningitis_KEGG_upreg_barplot}
EnrichKEGG_Upreg = enrichKEGG(Upreg2$entrezgene_id, organism = "hsa",keyType = "ncbi-geneid", qvalueCutoff = 0.05, use_internal_data = TRUE)

barplot(EnrichKEGG_Upreg, showCategory = 30)
```

```{r meningitis_KEGG_dotplot}

dotplot(EnrichKEGG, x="GeneRatio", showCategory = 30, font.size = 8, color="qvalue", orderBy ="x", title = "KEGG")

#dotplot(EnrichMKEGG_Upreg, x="GeneRatio", showCategory = 40, font.size = 8, color="qvalue", orderBy ="x", title = "KEGG module")
```

Upreg

```{r meningitis_KEGG_upreg_dotplot}
dotplot(EnrichKEGG_Upreg, x="GeneRatio", showCategory = 30, font.size = 8, color="qvalue", orderBy ="x", title = "KEGG")
```



```{r meningitis_GOterms_cnetplot}
## convert gene ID to Symbol
geneList_meningitis <- voom_tt2[,2]
names(geneList_meningitis) <- as.character(voom_tt2[,8])


GO_meningitis <- setReadable(Cluster_GO, 'org.Hs.eg.db', 'ENTREZID')


## categorySize can be scaled by 'pvalue' or 'geneNum'


cnetplot(GO_meningitis, foldChange=geneList_meningitis, categorySize="qvalue")


```

Upreg 

```{r meningitis_GOterms_Upreg_cnetplot }
GO_meningitis_Upreg_x <- setReadable(Cluster_upreg_GO, 'org.Hs.eg.db', 'ENTREZID')
cnetplot(GO_meningitis_Upreg_x, foldChange=geneList_meningitis, categorySize="qvalue")
```


```{r meningitis_GOterms_circular_cnetplot}
cnetplot(GO_meningitis_Upreg_x, foldChange=geneList_meningitis, circular = TRUE, colorEdge = TRUE)
```

Upreg

```{r meningitis_GOterms_Upreg_circular_cnetplot}
cnetplot(GO_meningitis, foldChange=geneList_meningitis, circular = TRUE, colorEdge = TRUE)

```

```{r meningitis_GOterms_cnetplot_category}
cnetplot(GO_meningitis, node_label="all")

```

Upreg

```{r meningitis_GOterms_Upreg_cnetplot_category}
cnetplot(GO_meningitis_Upreg_x, node_label="all")
```


```{r meningitis_GOterms_pathway_heatplot}
heatplot(GO_meningitis_Upreg_x, foldChange=geneList_meningitis)+ theme(axis.text.x = element_text(size = 4))+ ggtitle("GO pathway\nmeningitis CSF")+ theme (plot.title = element_text(size=12, face = "bold"))+ guides(fill=guide_colorbar(title = "Fold change"))
```

Upreg 

```{r meningitis_GOterms_Upreg_pathway_heatplot}
heatplot(GO_meningitis, foldChange=geneList_meningitis)+ theme(axis.text.x = element_text(size = 4))+ ggtitle("GO pathway\nmeningitis CSF")+ theme (plot.title = element_text(size=12, face = "bold"))+ guides(fill=guide_colorbar(title = "Fold change"))

```

```{r meningitis_GOterms_emapplot}
emapplot(GO_meningitis_Upreg_x)
```

Upreg

```{r meningitis_GOterms_Upreg_emapplot}
emapplot(GO_meningitis)

```


```{r meningitis_KEGGterms_cnetplot}
KEGG_meningitis <- setReadable(EnrichKEGG, 'org.Hs.eg.db', 'ENTREZID')

cnetplot(KEGG_meningitis, foldChange=geneList_meningitis, cateKEGGrySize="qvalue")

```

Upreg

```{r meningitis_KEGGterms_Upreg_cnetplot }
## convert gene ID to Symbol
geneList_meningitis <- voom_tt2[,2]
names(geneList_meningitis) <- as.character(voom_tt2[,8])

KEGG_meningitis_Upreg_x <- setReadable(EnrichKEGG_Upreg, 'org.Hs.eg.db', 'ENTREZID')

## cateKEGGrySize can be scaled by 'pvalue' or 'geneNum'
cnetplot(KEGG_meningitis_Upreg_x, foldChange=geneList_meningitis, cateKEGGrySize="qvalue")
```

```{r meningitis_KEGGterms_circular_cnetplot}
cnetplot(KEGG_meningitis, foldChange=geneList_meningitis, circular = TRUE, colorEdge = TRUE)

```

Upreg

```{r meningitis_KEGGterms_Upreg_circular_cnetplot}
cnetplot(KEGG_meningitis_Upreg_x, foldChange=geneList_meningitis, circular = TRUE, colorEdge = TRUE)
```

```{r meningitis_KEGGterms_cnetplot_category}
cnetplot(KEGG_meningitis, node_label="category")

```

Upreg

```{r meningitis_KEGGterms_Upreg_cnetplot_category}
cnetplot(KEGG_meningitis_Upreg_x, node_label="category")
```

```{R meningitis_KEGGterms_pathway_heatplot}
heatplot(KEGG_meningitis, foldChange=geneList_meningitis)+ theme(axis.text.x = element_text(size = 4))+ ggtitle("KEGG pathway\nmeningitis CSF high throughput")+ theme (plot.title = element_text(size=12, face = "bold"))+ guides(fill=guide_colorbar(title = "Fold change"))

```

Upreg

```{r meningitis_KEGGterms_Upreg_pathway_heatplot}
heatplot(KEGG_meningitis_Upreg_x, foldChange=geneList_meningitis)+ theme(axis.text.x = element_text(size = 4))+ ggtitle("KEGG pathway\nmeningitis CSF high throughput")+ theme (plot.title = element_text(size=12, face = "bold"))+ guides(fill=guide_colorbar(title = "Fold change"))
```

```{r meningitis_KEGGterms_emapplot}
emapplot(KEGG_meningitis)

```

Upreg

```{r meningitis_KEGGterms_Upreg_emapplot}
emapplot(KEGG_meningitis_Upreg_x)
```


